example-apps/internal-knowledge-search/app-ui/src/components/SearchApplicationSettings.tsx

```diff
@@ -1,23 +1,222 @@
+                                                            "must_not": {
+                                                                "exists": {
+                                                                    "field": "_allow_access_control"
+                                                                }
+                                                            }
+                                                        }
+                                                    },
+                                                    {
+                                                        "terms": {
+                                                            "_allow_access_control.enum": {{#toJson}}access_control{{/toJson}}
+                                                        }
+                                                    }
+                                                ]
+                                            }
+                                        }
+                                    }
+                                }`
+                            }
+                        }
+                    }
+                ],
+                "restriction": {
+                    "workflows": [
+                        "search_application_query"
+                    ]
+                }
+            }
+        }
+    }
+    const createPersonaAPIKey = async(persona) => {
+        const roleDescriptor = persona == "admin" ? defaultRoleDescriptor : await personaRoleDescriptor(persona)
+        const apiKeyPath = searchEndpoint + "/_security/api_key"
+        const response = await fetch(apiKeyPath, {
+            method: "POST",
+            headers: {
+                "Authorization": "Basic " + btoa(appUser + ":" + appPassword),
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify({
+                "name": appName+"-internal-knowledge-search-example-"+persona,
+                "expiration": "1h",
+                "role_descriptors": roleDescriptor,
+                "metadata": {
+                    "application": appName,
+                    "createdBy": appUser
+                }
+            })
+        })
+        const jsonData = await response.json()
+        return jsonData.encoded
+    }
+
+    const [searchPersonaOptions, setSearchPersonaOptions] = useState(["admin"]);
+
+    useEffect(()=>{
+        (async()=>{
+            const fetchedPersonas = await fetchPersonaOptions()
+            setSearchPersonaOptions(fetchedPersonas)
+            if (searchPersonaAPIKey == "missing") {
+                const createdAPIKey = await createPersonaAPIKey(searchPersona)
+                updateSearchPersonaAPIKey(createdAPIKey)
+            }
+        })()
+    },[])
+
+    const [isOpen, setIsOpen] = useState(false);
+
+    const toggleDropdown = () => {
+        setIsOpen(!isOpen);
+    };
+
+    const handlePersonaChange = async (value: string) => {
```

Comment:
> Instead of creating an API key at the same time as changing the persona, it'd be cleaner to update the Persona, and then have an effect that reacts to a Persona change to create a new API key.

I don't think I grok the distinction. Is there a way to "watch" persona for a change, other than to call something from it's update function? Like, is there an event that is triggered that I can listen for when a model changes?

> Even better would be to store a separate API key per persona in the state so you only ever create one API key per persona instead of recreating every time a persona changes back and forth.

Fair. I'd considered adding such a thing to the SearchApplicationSettings model, but felt like it was just getting too big. SearchPersonas probably needs to be its own model, and SearchApplicationSettings just references the ID/Name of a single persona at a time.

Reply:
> I don't think I grok the distinction. Is there a way to "watch" persona for a change, other than to call something from it's update function? Like, is there an event that is triggered that I can listen for when a model changes?

Yes! `useEffect`s dependency array allows you do this:

```
useEffect(async () => {
  const fetchApiKey = async () => { 
     return await createPersonaApiKey(value);
  }
  const apiKey = fetchApiKey();
  updateApiKey(apiKey);
}, [searchPersona]);
```

Note that here we do define an async function inside of this effect and then call it, because we care about the result of that async await call inside of this effect.

> Fair. I'd considered adding such a thing to the SearchApplicationSettings model, but felt like it was just getting too big. SearchPersonas probably needs to be its own model, and SearchApplicationSettings just references the ID/Name of a single persona at a time.

Yeah I figured and this is a fine compromise for an app like this.